Add SymInt to Scalar

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84958

This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure

eellison committed:
[WIP] Add SymInt to Scalar

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Add SymInt to Scalar"




[ghstack-poisoned]

eellison committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

Krovatkin commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -145,7 +160,8 @@ inline bool THPUtils_checkScalar(PyObject* obj) {
 #endif
-  return PyFloat_Check(obj) || PyLong_Check(obj) || PyComplex_Check(obj);
+  return PyFloat_Check(obj) || PyLong_Check(obj) ||
+      PyComplex_Check(obj) | torch::is_symint_node(py::handle(obj));
```
why are you using the bitwise or `|`?

Krovatkin commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -136,7 +150,8 @@ inline bool THPUtils_checkDouble(PyObject* obj) {
   }
 #endif
-  return PyFloat_Check(obj) || PyLong_Check(obj);
+  return PyFloat_Check(obj) ||
```
why are you using the bitwise or `|`?

Krovatkin commented on c10/core/Scalar.h:
```diff
@@ -158,14 +198,68 @@ class C10_API Scalar {
+    tag = Tag::HAS_i;
+  }
+
+  // We can't set v in the initializer list using the
```
a random comment, I think struct initialization only works if a struct doesn't have any c-tors :) the other compilers must be more lenient than msvc :)

Krovatkin commented on c10/core/Scalar.h:
```diff
@@ -30,12 +31,23 @@ namespace c10 {
   Scalar(type vv) : Scalar(vv, true) {}
 
+  void destroy() {
+    if (Tag::HAS_si == tag) {
```
I think we run into performance issues when we did something similar for `SymInt`. It might be less of an issue for Scalars. Tagging @swolchok

ezyang replied:
This shouldn't be an issue for scalars as we do not create and destroy large numbers of them, unlike tensors.

Krovatkin commented on c10/core/Scalar.h:
```diff
@@ -158,14 +198,68 @@ class C10_API Scalar {
+
+  C10_ALWAYS_INLINE void moveFrom(Scalar&& rhs) noexcept {
+    if (rhs.tag == Tag::HAS_si) {
+      new (&v.si) c10::SymInt(rhs.v.si);
```
We should be able to just do
```
v.si = std::move(rhs.v.si)
```
It's a little bit scary you need to call `release_` on SymInt manually

?

Krovatkin commented on c10/core/Scalar.h:
```diff
@@ -158,14 +198,68 @@ class C10_API Scalar {
   }
 
+  Scalar(Scalar&& rhs) noexcept : tag(rhs.tag) {
+    moveFrom(std::move(rhs));
```
okay, we don't need `destroy` call here since we are constructing a brand new `Scalar` here

Krovatkin commented on c10/core/Scalar.h:
```diff
@@ -158,14 +198,68 @@ class C10_API Scalar {
+    } u;
+    Payload() : u() {}
+    ~Payload() {}
+    c10::SymInt si;
```
it looks like we will always be storing a `SymIntNode` in a SymInt since if it's a concrete int, we would use `.u.i`. 
This means we will be incurring `is_symbolic` check every time we are destroying a `SymInt` again not sure if it matters for Scalars. 

@ezyang @swolchok

Krovatkin commented on c10/core/Scalar.h:
```diff
@@ -158,14 +198,68 @@ class C10_API Scalar {
+
+  Scalar(const Payload& p, Scalar::Tag t) : tag(t) {
+    if (t == Tag::HAS_si) {
+      v.si = c10::SymInt(p.si);
```
why can't we just do
``` 
v.si = p.si
```
?

Krovatkin commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -1461,6 +1461,18 @@ TEST(TestSymInt, TestIntrusive) {
 }
 
+TEST(TestSymInt, TestIntrusiveScalar) {
+  auto a = c10::make_intrusive<c10::SymIntNodeImpl>();
```
There seems to be a few different scenarios we might want to test
Axis 1 : copying from Scalar(int) into Scalar(SymInt) and vice versa
Axis 2: different types of copying: copy c-tor, assignment operator w/ , w/o move

Krovatkin reviewed (commented):
Added a few comments!

ezyang commented:
> for some reason torch.add(tensor, sym_int) got matched to the add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor schema

This is expected. The symint will turn into a wrapped number but we have enough logic to make this work already on master.

ezyang commented on c10/core/Scalar.h:
```diff
@@ -30,12 +31,23 @@ namespace c10 {
  * but don't know its type.
  */
 class C10_API Scalar {
+ public:
```
delete this

ezyang commented on c10/core/Scalar.h:
```diff
@@ -98,6 +117,27 @@ class C10_API Scalar {
+    } else {
+      this->v.si = other.v.si;
+    }
+    return *this;
```
These probably don't do self assignment correctly

eellison committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

ezyang commented on c10/core/Scalar.h:
```diff
@@ -49,21 +60,26 @@ class C10_API Scalar {
+    if (Tag::HAS_b == tag) {                                            \
+      return checked_convert<type, bool>(v.u.i, #type);                 \
+    } else {                                                            \
+      return checked_convert<type, int64_t>(v.u.i, #type);              \
```
Is it safe to fall through here? There is now one more new tag type

ezyang commented on c10/core/Scalar.h:
```diff
@@ -49,21 +60,26 @@ class C10_API Scalar {
+
+  SymInt toSymInt() const {
+    TORCH_CHECK(Tag::HAS_si == tag);
+    return v.si;
```
This should work if the Scalar contains an integer too.

ezyang commented on c10/core/Scalar.h:
```diff
@@ -86,9 +102,11 @@ class C10_API Scalar {
       "isIntegral is deprecated. Please use the overload with 'includeBool' parameter instead.")
   bool isIntegral() const {
+    // add symint here ?
     return Tag::HAS_i == tag;
```
yes, symints should report that they're integral

ezyang commented on c10/core/Scalar.h:
```diff
@@ -98,6 +116,24 @@ class C10_API Scalar {
   }
+  bool isSymInt() const {
+    return Tag::HAS_si == tag;
+  }
```
Not sure if this query makes sense.

ezyang commented on c10/core/Scalar.h:
```diff
@@ -108,12 +144,12 @@ class C10_API Scalar {
+      return v.u.i == num;
     } else {
       // boolean scalar does not equal to a non boolean value
       return false;
```
This seems no longer complete

ezyang reviewed (changes requested):
need a fine toothed comb through all of the preexisting methods

eellison committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

eellison committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

ezyang commented on c10/core/Scalar.h:
```diff
@@ -49,21 +60,32 @@ class C10_API Scalar {
+    } else if (Tag::HAS_i == tag) {                                     \
+      return checked_convert<type, int64_t>(v.u.i, #type);              \
+    } else if (Tag::HAS_si == tag) {                                    \
+      TORCH_CHECK(false)                                                \
```
It would be more informative for these to be `TORCH_INTERNAL_ASSERT` as they represent internal invariant violation in our code (and bugs should be filed)

ezyang commented on c10/core/Scalar.h:
```diff
@@ -74,7 +96,9 @@ class C10_API Scalar {
   template <typename T>
   T to() const = delete;
 
+  // audit uses of data_ptr
```
TODO

ezyang commented on c10/core/Scalar.h:
```diff
@@ -108,12 +156,14 @@ class C10_API Scalar {
     } else if (isIntegral(/*includeBool=*/false)) {
-      return v.i == num;
+      // test if symint or not
+      // guard on bool
```
TODO

ezyang commented:
This still doesn't seem done. I'm going to resuggest something I suggested earlier, which is to just store an IValue inside Scalar. If you can get the import cycles to work out, it will save you from a lot of delicate logic dealing with the nontrivial destructor.

eellison commented:
> This still doesn't seem done

Yea, it's not done sorry, I just pushed our comments in case it was super high pri for someone to take. have been spotty with bits of pto yesterday/monday but i can polish up and finish today.

ezyang commented:
I'll be taking this over

ezyang committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

ezyang committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

ezyang committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

ezyang committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

ezyang committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

ezyang committed:
Update on "Add SymInt to Scalar"


This is by no means comprehensive, but adds initial support for SymInt as a Scalar.

Things that don't work yet but need to:
- for some reason `torch.add(tensor, sym_int)` got matched to the `add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor` schema
- `x + sym_int` failed bc we tried to turn `x` into a sym int: 
```
              "__radd__",
              [](c10::SymIntNode a, py::object b) -> c10::SymIntNode {
                auto snb = toSymIntNode(a, b);
                return a->add(snb);
              })
 ```  
- Many more things I'm sure


[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
